server/src/Lib.hs

```diff
@@ -147,42 +152,65 @@ buildTriplet appDB trip@(BinaryTriplet bin pkg sys) = withApplicationDB appDB $
 
+-- | Write the data from the input 'Handle' to the output 'Handle' and collect
+-- the read bytes into a lazy bytestring.
+tee' :: Handle -> Handle -> IO Lazy.ByteString
+tee' input output = do
+  str <- Lazy.hGetContents input
+  result <-
+    BSL.fromChunks
+      <$> traverse (\chunk -> chunk <$ BS.hPut output chunk) (Lazy.toChunks str)
+  IO.hClose input
+  pure result
+
 nixBuild :: BinaryTriplet -> IO (Either String ByteString)
 nixBuild (BinaryTriplet bin pkg sys) = runExceptT $ do
-  -- TODO stream build process to stderr
   -- TODO make sure we can't get any XSS-like shenanigans
-  (exit, stdout, stderr) <-
-    liftIO $
-      Proc.readProcessWithExitCode
-        "nix"
-        ( ["build", "nixpkgs#legacyPackages." <> show sys <> ".pkgsStatic." <> unPackageName pkg]
-            -- Don't make a result symlink, just print it to stdout
-            <> ["--no-link", "--print-out-paths"]
-            -- Only allow access to files from NIX_PATH. Derivations can't read files like `/etc/shadow`.
-            <> ["--option", "restrict-eval", "true"]
-            -- Disallow IFD.  Probably not necessary, since Nixpkgs doesn't contain derivations that do IFD, but better to be on the safe side.
-            <> ["--option", "allow-import-from-derivation", "false"]
-            -- Make sure the sandbox is always used.
-            <> ["--option", "sandbox-fallback", "false"]
+  let process =
+        ( Proc.proc
+            "nix"
+            ( ["build", "nixpkgs#legacyPackages." <> show sys <> ".pkgsStatic." <> unPackageName pkg]
+                -- Don't make a result symlink, just print it to stdout
+                <> ["--no-link", "--print-out-paths"]
+                -- Print build logs to stderr
+                <> ["--print-build-logs"]
+                -- Only allow access to files from NIX_PATH. Derivations can't read files like `/etc/shadow`.
+                <> ["--option", "restrict-eval", "true"]
+                -- Disallow IFD.  Probably not necessary, since Nixpkgs doesn't contain derivations that do IFD, but better to be on the safe side.
+                <> ["--option", "allow-import-from-derivation", "false"]
+                -- Make sure the sandbox is always used.
+                <> ["--option", "sandbox-fallback", "false"]
+            )
         )
-        ""
+          { Proc.std_in = Proc.CreatePipe,
+            Proc.std_out = Proc.CreatePipe,
+            Proc.std_err = Proc.CreatePipe
+          }
+  (exit, stdout, stderr) <- ExceptT $
+    liftIO $
+      Proc.withCreateProcess process $ \minh mouth merrh processHandle -> do
+        case (minh, mouth, merrh) of
+          (Just inh, Just outh, Just errh) -> do
+            IO.hClose inh
+            (stdout, stderr) <- concurrently (tee' outh IO.stdout) (tee' errh IO.stderr)
+            exit <- Proc.waitForProcess processHandle
+            pure $ pure (exit, stdout, stderr)
+          _ -> pure $ throwError "couldn't read process output"
```

Comment:
Is this something you can capture in a separate function? Or combine with `tee'`?

Reply:
Sure, let's bikeshed about the name:

How about `streamProcessWithExitCode`? I think extracting everything except for the `tee'` is useless because that would essentially just yield the same API as that `withCreateProcess` already does modulo some pattern matching on the `Maybe Handle`s.